### ISSUE TITLE
Proposal: Add "develop" or "release" branch

### ISSUE THREAD

**Issue by lukastaegert** [rollup/rollup#2268]
Our current branching strategy of
* only having one central master branch and
* separate branches for improvements that are rebased against master before merging

is quite effective with regard to avoiding complex merges and difficult bisections. There are some drawbacks, however, that I hope to fix with my proposal:
* when a PR is merged, it is immediately marked as "merged" even though it is not released (requiring us to add labels to provide some clarification)
* if a user wants to check the code of the currently released version of rollup, they need to dig for release tags
* it does not allow us to have automatic deployments at some point while retaining the ability to collect feature PRs to have larger releases that include related features and can be announced "with more fanfare" e.g. on Twitter

# Proposal

* We introduce a new branch "develop" which for internal purposes serves the role of current master in that (at least internal) PRs are forked from this branch and merged back into this branch
* "master" serves the role of a release branch. When we do a release, the current "develop" branch is merged into master and then current master is published.
* In the future, this may even be done in an automatic way provided we take care of updating the changelog before that. At the moment each time a PR is merged to master, I add the changelog update to the merge commit. Something like this (or via a separat commit) could be done while merging into the "develop"  branch.
* Even though PRs are forked from "develop", they should still be raised agains "master". That way—unless we delete a PR branch prematurely—github should mark PRs as "merged" precisely when they merged into master and thus released
* If we want to do a "hotfix" i.e. we already merged new PRs into "develop" but want to release an important fix right away without releasing the features yet, we can directly merge a PR into master and then merge master back into "develop". This will make bisecting a little more difficult, though, so we should only do this for small fixes.

**Comment by guybedford** [comment]
How about merging into master as the development branch and then having a separate release branch for each major (which would currently be `0.x`)?

**Comment by lukastaegert** [comment]
Something along these lines should be easy to implement and even preserve the full history: https://stackoverflow.com/a/14470212

**Comment by lukastaegert** [comment]
Ok, thinking about this a little more, would you agree to the following approach:
* We leave master as it is
* A new branch "release" is added which represents what is currently released
* We change the pull request template to instruct people to raise their PRs agains the release branch
* Existing PRs are changed to be raised against release (should just be a simple matter of clicking around a little)

I think this should be really easy to understand also for external collaborators who will probably intuitively fork from master anyway. Would this work for you? Then I will create a PR for the PR template. And we can still discuss the rest.

**Comment by guybedford** [comment]
> We change the pull request template to instruct people to raise their PRs agains the release branch

Is the reasoning here that something should be released as soon as it is merged so that we are following continuous integration? And that larger features worked on by core maintainers would go into master because they represent more experimental architecture shifts?

**Comment by lukastaegert** [comment]
> Is the reasoning here that something should be released as soon as it is merged so that we are following continuous integration? And that larger features worked on by core maintainers would go into master because they represent more experimental architecture shifts?

Yes, this is it exactly. Setting up continuous integration would not be the first step but is definitely a goal I would hope to achieve at some point. As for the second point, smaller and more focused PRs are usually preferable especially with regard to reviews and having early feedback. But as for releases, certain features could make more sense when published together (i.e. a suit of plugin hooks; or when we want to collect several unrelated breaking changes to only have a single breaking release so that people who update only have to dig through the release notes of this particular release).

**Comment by lukastaegert** [comment]
The only reason I would not include continuous integration initially is the open question of how the changelog is managed. But this may be the topic of another issue once we get there.

**Comment by guybedford** [comment]
It sounds like the core of what is needed is a better release / merge / changelog / integration / announcement process.

Going back to some of the problems you mentioned - 

> when a PR is merged, it is immediately marked as "merged" even though it is not released (requiring us to add labels to provide some clarification)

This is a problem every single project on GitHub has. I don't think there is confusion by most people that merged and released mean two separate things. That is, I'm not sure I see this as a problem. Most OS projects I've worked with do things this way - closing issues when merged, and having releases be a separate process to issue resolution and merging.

> if a user wants to check the code of the currently released version of rollup, they need to dig for release tags

Having release branches per major can be useful here.

> it does not allow us to have automatic deployments at some point while retaining the ability to collect feature PRs to have larger releases that include related features and can be announced "with more fanfare" e.g. on Twitter

One thing that might help here is standardised commit messages on merge:

* On merge each PR is squashed into a single commit explicitly stating the feature issue number, PR number etc. eg `"treeshaking: fixes #1111 allowing undefined literal value handling`
* Release management is then simplified as individual commits can now be easily cherry-picked between releases making backporting simpler etc
* Changelog generation is then simply listing the commits from the last release, and automated processes can then use this listing to generate that as well as release announcements.

My worry with the existing proposal is that introducing a separate branch from master for development may result in new contributors creating PRs against the wrong branch. By not following a common pattern this can create friction for new maintainers and contributors.

> If we want to do a "hotfix" i.e. we already merged new PRs into "develop" but want to release an important fix right away without releasing the features yet, we can directly merge a PR into master and then merge master back into "develop". This will make bisecting a little more difficult, though, so we should only do this for small fixes.

This kind of cross-branch merging could be done more easily with the single-commit approach.

**Comment by lukastaegert** [comment]
Sounds like some good ideas! So with your permission, I will just tweak our setup very slightly:
* Create a `release` branch for the currently released version

Beyond that, people should still just fork their PRs from master and merge them back into master. Beyond that, we might experiment with manually switching the merge target of PRs in GitHub to the release branch to see if this would work to trick GitHub adding the label at the right time or just leave it as it is as the goal is to make more and smaller releases in the future anyway.

I like the idea of squashing and using a specific commit message format to handle automatically writing the changelog as that way, we would not require external contributors to adhere to a strict commit message format. It should also motivate us to keep our PRs small and focused. I hope we can set up some automation to enforce the commit message format specifically for merges to master and then set up continuous deployment in a separate step.

**Comment by lukastaegert** [comment]
As soon as we have PRs with breaking changes, we could then set up a second release branch for the next major which is merged in the current release branch and removed once this major is released.

**Comment by guybedford** [comment]
> Create a release branch for the currently released version

Sure, that sounds sensible to me.

Hopefully we can make the 1.0 as backwards compatible as possible so that we won't need to continue maintaining a 0.x branch for bug fixes etc. But if that ever changes in future we can reconsider then.

> I like the idea of squashing and using a specific commit message format to handle automatically writing the changelog as that way, we would not require external contributors to adhere to a strict commit message format. It should also motivate us to keep our PRs small and focused. I hope we can set up some automation to enforce the commit message format specifically for merges to master and then set up continuous deployment in a separate step.

Definitely happy to move in these directions, but we just need to be careful about the message format to ensure we have useful information.

**Comment by lukastaegert** [comment]
I added a release branch now with the sole purpose of reflecting the current release and the option to add automation here later.

I also played around with the configuration of the master branch on Github and did the following changes which provide a very nice workflow:
* On the web interface, "Squash and merge" is now the only allowed merge method; in fact, "Squash and merge" turned out to work more like "Squash and rebase" in that it creates a very nice history.
* I also required that a branch can only be merged if it is up to date with master and all checks pass. This has the effect that if a branch is not up to date with master, an "update" button will appear that automatically merges master into that branch and runs the checks again. Even though this temporarily creates a branched history, squash and merge will make this branching vanish again (as I just tried out). So as long as PRs are only merged through the web interface, it is now safe to just merge master into a PR instead of rebasing the PR onto master without tainting the history which should make this a lot easier.
